runtime/complex.go

```diff
@@ -130,3 +134,130 @@ func complexCoerce(o *Object) (complex128, bool) {
 	}
 	return complex(floatO, 0.0), true
 }
+
+func complexNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	if argc == 0 {
+		return newObject(t), nil
+	}
+	if argc > 2 {
+		return nil, f.RaiseType(TypeErrorType, "'__new__' of 'complex' requires at most 2 arguments")
+	}
+	if t != ComplexType {
+		// Allocate a plain complex then copy it's value into an object
+		// of the complex subtype.
+		x, raised := complexNew(f, ComplexType, args, nil)
+		if raised != nil {
+			return nil, raised
+		}
+		result := toComplexUnsafe(newObject(t))
+		result.value = toComplexUnsafe(x).Value()
+		return result.ToObject(), nil
+	}
+	o := args[0]
+	if argc == 1 {
+		if complexSlot := o.typ.slots.Complex; complexSlot != nil {
+			result, raised := complexSlot.Fn(f, o)
+			if raised != nil {
+				return nil, raised
+			}
+			if !result.isInstance(ComplexType) {
+				format := "__complex__ returned non-complex (type %s)"
+				return nil, f.RaiseType(TypeErrorType, fmt.Sprintf(format, result.typ.Name()))
+			}
+			return result, nil
+		}
+		if floatSlot := o.typ.slots.Float; floatSlot != nil {
+			result, raised := floatConvert(floatSlot, f, o)
+			if raised != nil {
+				return nil, raised
+			}
+			f := toFloatUnsafe(result).Value()
+			return NewComplex(complex(f, 0)).ToObject(), nil
+		}
+		if !o.isInstance(StrType) {
+			return nil, f.RaiseType(TypeErrorType, "complex() argument must be a string or a number")
+		}
+		s := toStrUnsafe(o).Value()
+		result, err := parseComplex(s)
+		if err != nil {
+			return nil, f.RaiseType(ValueErrorType, "complex() arg is a malformed string")
+		}
+		return NewComplex(result).ToObject(), nil
+	}
+
+	// TODO: fix me
+	return nil, nil
+}
+
+// ParseComplex converts the string s to a complex number.
+// If string is well-formed (one of these forms: <float>, <float>j,
+// <float><signed-float>j, <float><sign>j, <sign>j or j, where <float> is
+// any numeric string that's acceptable by strconv.ParseFloat(s, 64)),
+// ParseComplex returns the respective complex128 number.
+func parseComplex(s string) (complex128, error) {
+	c := strings.Count(s, "(")
+	if (c > 1) || (c == 1 && strings.Count(s, ")") != 1) {
+		return complex(0, 0), errors.New("Malformed complex string, more than one matching parantheses")
+	}
+	ts := strings.Trim(s, "() ")
+	re := `(?i)(?:(?:(?:(?:\d*\.\d+)|(?:\d+\.?))(?:[Ee][+-]?\d+)?)|(?:infinity)|(?:nan)|(?:inf))`
+	fre := `[-+]?` + re
+	sre := `[-+]` + re
+	fsfj := `(?:(?P<real1>` + fre + `)(?P<imag1>` + sre + `)j)`
+	fsj := `(?:(?P<real2>` + fre + `)(?P<sign2>[-+])j)`
+	fj := `(?P<imag3>` + fre + `)j`
+	f := `(?P<real4>` + fre + `)`
+	sj := `(?P<sign5>[-+])j`
+	j := `(?P<onlyJ>j)`
+	r := regexp.MustCompile(`^(?:` + fsfj + `|` + fsj + `|` + fj + `|` + f + `|` + sj + `|` + j + `)$`)
+	subs := r.FindStringSubmatch(ts)
+	if subs == nil {
+		return complex(0, 0), errors.New("Malformed complex string, no mathing pattern found")
+	}
+	res := make(map[string]string)
```

Comment:
Allocation in Go is relatively expensive. I think you can avoid allocating this map by using indices for the sub-matches and using a fixed length array: var res [6]string. Constants for the index names would make it more readable.